src/i2c.rs

```diff
@@ -718,6 +1077,536 @@ impl<I2C: Instance> I2c<I2C> {
+    fn restart_transfer(&mut self, expected_bytes: usize) {
+        self.i2c
+            .cr2()
+            .modify(|_, w| w.nbytes().set(expected_bytes as u8));
+    }
+
+    /// Manage checking for events in manual acking mode
+    fn get_target_event(&mut self) -> Result<Option<TargetEvent>, Error> {
+        let isr = self.read_isr_and_check_errors()?;
+
+        if isr.addr().is_match() {
+            if isr.txe().is_not_empty() {
+                // Flush the contents of TXDR without writing it to the bus. Doing so ensures that
+                // spurious data is not written to the bus and that the clock remains stretched when
+                // the ADDR flag is cleared and until data is written to TXDR (in the case of a
+                // target write operation)
+                self.i2c.isr().write(|w| w.txe().set_bit());
+            }
+            // Reset SBC, reload, and nbytes. They're used differently for a read or write, but
+            // make sure they're in a known state before starting the next operation.
+            self.i2c.cr1().modify(|_, w| w.sbc().disabled());
+            self.i2c
+                .cr2()
+                .modify(|_, w| w.reload().completed().nbytes().set(0));
+
+            let address = isr.addcode().bits();
+
+            if isr.dir().is_read() {
+                self.i2c.icr().write(|w| w.addrcf().clear());
+                // TODO: Translate address into 10-bit stored address if relevant
+                Ok(Some(TargetEvent::TargetWrite {
+                    address: address as u16,
+                }))
+            } else {
+                // Manual ACK control uses slave byte control mode when reading data from the
+                // controller, so set it up here.
+                self.i2c.cr1().modify(|_, w| w.sbc().enabled());
+                self.i2c.cr2().modify(|_, w| w.reload().not_completed());
+                Ok(Some(TargetEvent::TargetRead {
+                    address: address as u16,
+                }))
+            }
+        } else if isr.stopf().is_stop() {
+            self.i2c.icr().write(|w| w.stopcf().clear());
+            Ok(Some(TargetEvent::Stop))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Block operation to wait for this device to be addressed or for a
+    /// transaction to be stopped. When the device is addressed, a
+    /// TargetEvent::Read or TargetEvent::Write event will be returned.
+    /// If the transaction was stopped, TargetEvent::Stop will be returned.
+    ///
+    /// If an error occurs while waiting for an event, this will be returned.
+    ///
+    /// Note: The TargetEvent type that is indicated is indicated from the
+    /// perspective of the Bus controller, so if a Read event is received,
+    /// then the Target should respond with a write, and vice versa.
+    pub fn wait_for_event(&mut self) -> Result<TargetEvent, Error> {
+        loop {
+            if let Some(event) = self.get_target_event()? {
```

Comment:
Can this be put in the base trait implementation as a default? It seems to be copied across the impls